design/progress-tracking.md

```diff
@@ -0,0 +1,244 @@
+
+When an actionset triggers a long-running task like the `CopyVolumeData`
+function, the only way to determine any sort of progress is to gain direct access
+to the logs and events of the task pod. This may not always be feasible as the
+user may not have the appropriate RBAC permissions to access these subresource
+endpoints in the pod's namespace. If the export operation takes a long time,
+it's also possible that the pod might be terminated prematurely without leaving
+behind any traces of logs to indicate how far along the task was.
+
+Being able to persist progress data will be very helpful for both live reporting
+of task progress, as well as future retrospection (e.g., the latency of every
+phase of an action).
+
+## Proposed Solution
+
+### Summary
+
+The `ActionSet` CRD's  `status` subresource will be updated with new fields to
+communicate the blueprint and phases' progress status to the user. The
+`kube.ExecOutput()` and `kube.Task()` interfaces will be updated to accept a new
+progress I/O writer.
+
+## Assumptions And Constraints
+
+* The progress computation should not compromise the main data protection task's
+latency nor lead to resource contention. Progress computation will be performed
+on a best-effort basis, where it may be de-prioritized with no guarantee on the
+accuracy of its result, or skipped entirely in the event of resource contention.
+
+### Changes To API
+
+The blueprint's overall progress will be reported under the new
+`status.progress` field. The progress of each phase will be included in the
+phase's subsection as `status.actions[*].phases[*].progess`.
+
+For example,
+
+```yaml
+status:
+  progress:
+    percentCompleted: 50 # 1 out of 2 phases are completed
+    lastTransitionTime: 2022-04-06 14:23:34
+  actions:
+  - blueprint: my-blueprint
+    name: action-00
+    phases:
+    - name: echo
+      state: completed
+      progress:
+        percentCompleted: 100
+        lastTransitionTime: 2022-04-06 14:13:00
+    - name: echo
+      state: completed
+      progress:
+        percentCompleted: 100
+        lastTransitionTime: 2022-04-06 14:23:34
+    name: action-01
+    phases:
+    - name: echo
+      state: pending
+      progress:
+        percentCompleted: 30
+        lastTransitionTime: 2022-04-06 14:30:31
+  state: pending
+```
+
+### Progress Tracking
+
+Since progress tracking may not be meaningful for short-lived tasks, we will
+limit the initial implementation effort to the following Kanister Functions
+which normally are used to invoke long-running operations:
+
+* `BackupData`
+* `BackupDataAll`
+* `RestoreData`
+* `RestoreDataAll`
+* `CopyVolumeData`
+* `CreateRDSSnapshot`
+* `ExportRDSSnapshotToLocation`
+* `RestoreRDSSnapshot`
+
+#### Blueprint Progress
+
+Initially, the blueprint overall progress is computed by checking the number of
+completed phases against the total number of phases:
+
+```
+blueprint_progress_percent = num_completed_phases / total_num_phases * 100
```

Comment:
should this be `action_progress_percent`?

Reply:
@viveksinghggits yes, good catch! To get the blueprint overall progress, we should be assessing the completed actions, not phases.